## å­Ÿå¾·å°”åï¼šéªŒè¯é›†ã€åŒå‘å­Ÿå¾·å°”ã€å…±å®šä½åˆ†æ

library(tidyverse)

load('exposure_dat.Rdata')

gene=read.table('./OR.txt',sep = '\t',header = T)
gene=gene$exposure

exposure_dat=exposure_dat[exposure_dat$exposure %in% gene,]

### éªŒè¯é›?----

#æå–ç»“å±€æ•°æ®,æ¢ä¸€ä¸ªé˜Ÿåˆ?
library(TwoSampleMR)
outcome_dat <- extract_outcome_data(snps=exposure_dat$SNP, outcomes="ebi-a-GCST003156")

# å–äº¤é›?
exposure_dat=exposure_dat[exposure_dat$SNP %in% outcome_dat$SNP,]


harmonised_dat <- harmonise_data(exposure_dat, outcome_dat)



## MR

mr_modified <- function(dat = harmonised_dat, prop_var_explained = T)
{
  mr_res <- mr(dat)
  
  pve <- dat %>% 
    dplyr::select(id.exposure, beta.exposure, se.exposure, samplesize.exposure) %>% 
    dplyr::group_by(id.exposure) %>% 
    dplyr::summarise(pve = sum((beta.exposure^2)/(beta.exposure^2 + samplesize.exposure*se.exposure^2)))
  
  if(prop_var_explained)
  {
    mr_res <- mr_res %>% 
      dplyr::left_join(pve, by = "id.exposure")
  }
  
  return(mr_res)
}




mr_res_vali <- mr_modified(harmonised_dat, prop_var_explained = T)

save(mr_res_vali,harmonised_dat,file ='mr_input_res_vali.Rdata')


result_or=generate_odds_ratios(mr_res_vali)
write.table(result_or[,4:ncol(result_or)],"OR_vali.txt",row.names = F,sep = "\t",quote = F)

#å°†ç»Ÿè®¡ç»“æœç»˜åˆ¶æ£®æ—å›¾


library(grid)
library(forestploter)


mydata=read.table("OR_vali.txt",header = T,sep = "\t")
mydata$` ` <- paste(rep(" ", 20), collapse = " ")
mydata$`OR (95% CI)` <- ifelse(is.na(mydata$or), "",sprintf("%.4f (%.4f - %.4f)",
                                                            mydata$or, mydata$or_lci95, 
                                                            mydata$or_uci95))
forest(mydata[,c(1:3,6,13,14)],
       est = mydata$or,
       lower =mydata$or_lci95, 
       upper = mydata$or_uci95,
       sizes =0.3,
       ci_column =5 ,
       ref_line = 1,
       xlim = c(0.05, 3),
)


## å…¶ä»–å¯è§†åŒ–ï¼ˆä¸€èˆ¬ä¸å¤ªå¥½ç”¨ï¼Œå› ä¸ºä¸€ä¸ªåŸºå› åªæœ‰å¾ˆæœ‰é™çš„eqtl
mr_res1=mr_res_vali[mr_res_vali$exposure=='APOBEC3G',]
harmonised_dat1=harmonised_dat[harmonised_dat$exposure=='APOBEC3G',]

#ç»˜åˆ¶æ•£ç‚¹å›?
mr_scatter_plot(mr_res1, harmonised_dat1)

#æ£®æ—å›?
res_single=mr_singlesnp(harmonised_dat1)
mr_forest_plot(res_single)

#æ¼æ–—å›?
mr_funnel_plot(singlesnp_results = res_single)

#ç•™ä¸€æ³•æ•æ„Ÿæ€§åˆ†æ?
mr_leaveoneout_plot(leaveoneout_results = mr_leaveoneout(harmonised_dat1))



### åå‘å­Ÿå¾·å°?
# æš´éœ²
bimr_SLE<- extract_instruments('ebi-a-GCST90011866', p1=5e-08, clump=TRUE)

outcome_gene<- extract_outcome_data(snps=bimr_SLE$SNP, outcomes="eqtl-a-ENSG00000239713")

bimr_SLE =bimr_SLE [bimr_SLE $SNP %in% outcome_gene$SNP,]


harmonised_SLE_gene <- harmonise_data(bimr_SLE, outcome_gene)

bimr_mr_SLE_gene <- mr(harmonised_SLE_gene)

result_or=generate_odds_ratios(bimr_mr_SLE_gene)
write.table(result_or[,4:ncol(result_or)],"bi_OR.txt",row.names = F,sep = "\t",quote = F)

#å°†ç»Ÿè®¡ç»“æœç»˜åˆ¶æ£®æ—å›¾


library(grid)
library(forestploter)


mydata=read.table("bi_OR.txt",header = T,sep = "\t")
## !!
mydata$outcome='APOBEC3G'
mydata$` ` <- paste(rep(" ", 20), collapse = " ")
mydata$`OR (95% CI)` <- ifelse(is.na(mydata$or), "",sprintf("%.4f (%.4f - %.4f)",
                                                            mydata$or, mydata$or_lci95, 
                                                            mydata$or_uci95))
forest(mydata[,c(1:3,6,12,13,14)],
       est = mydata$or,
       lower =mydata$or_lci95, 
       upper = mydata$or_uci95,
       sizes =0.3,
       ci_column =6 ,
       ref_line = 1,
       xlim = c(0.05, 3),
)



### å…±å®šä½åˆ†æ?----

load('mr_input_res.Rdata')

#å¦‚æœè¡¨å‹æ˜¯äºŒåˆ†ç±»å˜é‡ï¼ˆcaseå’Œcontrolï¼‰ï¼Œè¾“å…¥æ–‡ä»¶äºŒé€‰ä¸€ï¼?
#1ï¼‰rsç¼–å·`rs_id`ã€På€¼`pval_nominal`ã€SNPçš„æ•ˆåº”å€¼`beta`ã€æ•ˆåº”å€¼æ–¹å·®`varbeta`ï¼›ï¼ˆæ¨èï¼?
#2ï¼‰rsç¼–å·`rs_id`ã€På€¼`pval_nominal`ã€caseåœ¨æ‰€æœ‰æ ·æœ¬ä¸­çš„æ¯”ä¾‹`s`ï¼ŒMAFä¹Ÿè¦ï¼Œå†™åœ¨listæœ€å?

#å¦‚æœè¡¨å‹æ˜¯è¿ç»­å‹å˜é‡ï¼Œè¾“å…¥æ–‡ä»¶ä¸‰é€‰ä¸€ï¼?
#1ï¼‰rsç¼–å·`rs_id`ã€På€¼`pval_nominal`ã€è¡¨å‹çš„æ ‡å‡†å·®`sdY`ï¼?
#2ï¼‰rsç¼–å·`rs_id`ã€På€¼`pval_nominal`ã€æ•ˆåº”å€¼`beta`,æ•ˆåº”å€¼æ–¹å·? `varbeta`, æ ·æœ¬é‡`N`,æ¬¡ç­‰ä½åŸºå› é¢‘ç? `MAF`ï¼?
#3ï¼‰rsç¼–å·`rs_id`ã€På€¼`pval_nominal`ã€æ¬¡ç­‰ä½åŸºå› é¢‘ç‡ `MAF`ï¼?(æ¨è)


## ä¸‹è½½
data <- vcfR::read.vcfR("./eqtl-a-ENSG00000239713.vcf/eqtl-a-ENSG00000239713.vcf")

# 1.SNP ID rså¼€å¤?
# 2.effect alleleï¼Œæ­¤å¤„ç›¸å½“äºALT
# 3.other alleleï¼Œæ­¤å¤„ç›¸å½“äºREF
# 4.beta
# 5.se
# 6.pval


#æ•´ç†æ•°æ®
## å¤„ç†gt
gt=data@gt
gt=as.data.frame(gt)

colnames(gt)
gt$FORMAT[1]

library(tidyverse)

gt$`eqtl-a-ENSG00000239713`[1]

##!!!
gt=separate(gt,col='eqtl-a-ENSG00000239713',into = c('ES', 'SE',
                                       'LP','AF','SS',
                                       'ID'),sep = '\\:')

gc()



gt=na.omit(gt)
colnames(gt)=c('format','beta','se','logpvalue','eaf','samplesize','snp')
gt$beta=as.numeric(gt$beta)
gt$se=as.numeric(gt$se)
gt$logpvalue=as.numeric(gt$logpvalue)
gt$eaf=as.numeric(gt$eaf)
gt$samplesize=as.numeric(gt$samplesize)

gc()
gt$format=NULL


fix=data@fix
fix=as.data.frame(fix)
colnames(fix)
colnames(fix)=c('chr','pos','snp','ref','alt')
fix=fix[,1:5]


## åˆå¹¶gt fix
eqtl=left_join(fix,gt,by='snp')
eqtl=na.omit(eqtl)
## æŸ¥æ‰¾æŸ“è‰²ä½“å’Œä½ç½®
#22
#39605007

#MAFsä»£è¡¨æ¬¡è¦ç­‰ä½åŸºå› çš„é¢‘ç‡ï¼Œå› æ­¤å…¶èŒƒå›´åœ¨0åˆ?0.5ä¹‹é—´ã€‚å€¼å¤§äº?0.5çš„æ‰€æœ‰EAFé€šè¿‡ä»?1ä¸­å‡å»å®ƒä»¬è€Œè½¬åŒ–ä¸ºMAFã€?
eqtl$maf = ifelse(eqtl$eaf < 0.5, 
                  eqtl$eaf,
                     1 - eqtl$eaf)
eqtl$eaf=NULL


#	
#22
#39605007

eqtl=eqtl[eqtl$chr==22,]
eqtl$logpvalue=as.numeric(eqtl$logpvalue)
eqtl$p_value=10^(-eqtl$logpvalue)

eqtl$pos=as.numeric(eqtl$pos)

## ä¸Šä¸‹1mkb
eqtl=eqtl[eqtl$pos > 39605007-1000000 ,]
eqtl=eqtl[eqtl$pos < 39605007+1000000 ,]

my_eqtl=eqtl[,c('snp','p_value','maf')]

colnames(my_eqtl)=c('snp','pvalues','MAF')
my_eqtl=na.omit(my_eqtl)

my_eqtl=my_eqtl[my_eqtl$MAF>0 ,]

##-----æ¥ä¸‹æ¥gwasç–¾ç—…


library(TwoSampleMR)
## ï¼ï¼ï¼ï¼ï¼?
coloc_SLE_dat <- extract_outcome_data(snps = c(my_eqtl$snp),
                                     outcomes = "ebi-a-GCST90011866",
                                     proxies = F) %>% 
  mutate(chr.outcome = as.numeric(chr),
         pos.outcome = as.numeric(pos),
         outcome = "Systemic lupus erythematosus", 
         id.outcome = "ebi-a-GCST90011866")


##gwas:ç–¾ç—…
gwas=coloc_SLE_dat
gwas$beta=as.numeric(gwas$beta.outcome)
gwas$se=as.numeric(gwas$se.outcome)
#X=gwas$beta/gwas$se
#P=2*pnorm(q=abs(X), lower.tail=FALSE) #Z=1.96  P=0.05
#gwas$pvalue=P
## check Lp=log10(P)
gwas$varbeta=(gwas$se)^2

gwas=gwas[,c('SNP','pval.outcome',"beta",'varbeta')]

colnames(gwas)=c('snp','pvalues','beta','varbeta')

gwas=na.omit(gwas)

##---å¼€å§‹coloc
library(coloc)



input <- merge(my_eqtl, gwas, by="snp", all=FALSE, suffixes=c("_eqtl","_gwas"))


head(input)
library(coloc)

# å¦‚æŠ¥é”™ï¼Œå»é™¤é‡å¤å€?
#input=input[!duplicated(input$snp),]

coloc.abf(dataset1=list(pvalues=input$pvalues_gwas, type="cc", s=0.33, N=12653),
                    dataset2=list(pvalues=input$pvalues_eqtl, type="quant", N=30684),
                    MAF=input$MAF)

## æˆ–è€…ç”¨ä¸‹é¢çš„æ–¹æ³?

#my_eqtl=as.list(my_eqtl)
#my_eqtl[['type']]='quant'
#my_eqtl[['N']]=31430

#gwas=as.list(gwas)
#gwas[['type']]='cc'
#gwas[['N']]=12653
#gwas$pvalues=NULL
#coloc.abf(dataset1 = gwas,dataset2 = my_eqtl)


## è§£è¯»------
#H0ï¼šè¯¥åŒºåŸŸçš„ä¸¤ä¸ªæ€§çŠ¶éƒ½æ²¡æœ‰é—ä¼ å…³è?
#H1 / H2ï¼šåªæœ‰è¡¨å?1æˆ–è¡¨å?2åœ¨è¯¥åŒºåŸŸå…·æœ‰é—ä¼ å…³è”
#H3ï¼šä¸¤ä¸ªç‰¹å¾éƒ½ç›¸å…³ï¼Œä½†å› æœå˜é‡ä¸åŒ
#H4ï¼šä¸¤ä¸ªç‰¹å¾éƒ½ç›¸å…³å¹¶ä¸”å…±äº«ä¸€ä¸ªå› æœå˜é‡?

